src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionDescriptorCollectionProvider.cs

```diff
@@ -2,30 +2,44 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
 using System.Collections.ObjectModel;
 using System.Linq;
+using System.Threading;
 using Microsoft.AspNetCore.Mvc.Abstractions;
 using Microsoft.AspNetCore.Mvc.Infrastructure;
-using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Primitives;
 
 namespace Microsoft.AspNetCore.Mvc.Internal
 {
     /// <summary>
     /// Default implementation of <see cref="IActionDescriptorCollectionProvider"/>.
-    /// This implementation caches the results at first call, and is not responsible for updates.
     /// </summary>
-    public class ActionDescriptorCollectionProvider : IActionDescriptorCollectionProvider
+    public class ActionDescriptorCollectionProvider : IActionDescriptorCollectionProvider, IDisposable
     {
-        private readonly IServiceProvider _serviceProvider;
+        private readonly IActionDescriptorProvider[] _actionDescriptorProviders;
+        private readonly List<IDisposable> _disposables;
         private ActionDescriptorCollection _collection;
+        private int _version = -1;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="ActionDescriptorCollectionProvider" /> class.
         /// </summary>
-        /// <param name="serviceProvider">The application IServiceProvider.</param>
-        public ActionDescriptorCollectionProvider(IServiceProvider serviceProvider)
+        /// <param name="actionDescriptorProviders">The sequence of <see cref="IActionDescriptorProvider"/>.</param>
+        /// <param name="actionDescriptorChangeProviders">The sequence of <see cref="IActionDescriptorChangeProvider"/>.</param>
+        public ActionDescriptorCollectionProvider(
+            IEnumerable<IActionDescriptorProvider> actionDescriptorProviders,
+            IEnumerable<IActionDescriptorChangeProvider> actionDescriptorChangeProviders)
         {
-            _serviceProvider = serviceProvider;
+            _actionDescriptorProviders = actionDescriptorProviders
+                .OrderBy(p => p.Order)
+                .ToArray();
+
+            _disposables = new List<IDisposable>();
+            foreach (var changeTokenProvider in actionDescriptorChangeProviders)
+            {
+                _disposables.Add(ChangeToken.OnChange(changeTokenProvider.GetChangeToken, OnChange));
+            }
```

Comment:
Rather than registering here, should we poll the change tokens? The other alternative is to use a weak-reference. I really don't want to see disposable here.